src/url-loader/fs-url-resolver.ts

```diff
@@ -0,0 +1,142 @@
+/**
+ * @license
+ * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import * as pathlib from 'path';
+import {posix as posix} from 'path';
+import {format as urlLibFormat, Url} from 'url';
+import Uri from 'vscode-uri';
+
+import {parseUrl} from '../core/utils';
+import {FileRelativeUrl, PackageRelativeUrl, ScannedImport} from '../index';
+import {ResolvedUrl} from '../model/url';
+
+import {UrlResolver} from './url-resolver';
+
+const isWindows = process.platform === 'win32';
+
+/**
+ * Resolves package-relative urls to a configured root directory.
+ */
+export class FsUrlResolver extends UrlResolver {
+  protected readonly packageDir: string;
+  protected readonly packageUrl: ResolvedUrl;
+  constructor(
+      packageDir: string|undefined, private readonly hostname?: string) {
+    super();
+    this.packageDir =
+        normalizeFsPath(pathlib.resolve(packageDir || process.cwd()));
+    this.packageUrl =
+        this.brandAsResolved(Uri.file(this.packageDir).toString());
+    if (!this.packageUrl.endsWith('/')) {
+      this.packageUrl = this.brandAsResolved(this.packageUrl + '/');
+    }
+  }
+
+  resolve(
+      unresolvedHref: FileRelativeUrl|PackageRelativeUrl,
+      baseUrl: ResolvedUrl = this.packageUrl,
+      _import?: ScannedImport): ResolvedUrl|undefined {
+    const resolvedHref = this.simpleUrlResolve(unresolvedHref, baseUrl);
+    if (resolvedHref === undefined) {
+      return undefined;
+    }
+    const url = parseUrl(resolvedHref);
+    if (this.shouldHandleAsFileUrl(url)) {
+      return this.handleFileUrl(url, unresolvedHref);
+    }
+    return this.brandAsResolved(resolvedHref);
+  }
+
+  protected shouldHandleAsFileUrl(url: Url) {
+    if (this.hostname !== undefined && url.hostname === this.hostname) {
+      return true;
+    }
+    return url.protocol === 'file:' && (!url.host || url.host === 'localhost');
+  }
+
+  /**
+   * Take the given URL which is either a file:// url or a url with the
+   * configured hostname, and treat its pathname as though it points to a file
+   * on the local filesystem, producing a file:/// url.
+   *
+   * Also corrects sibling URLs like `../foo` to point to
+   * `./${component_dir}/foo`
+   */
+  private handleFileUrl(url: Url, unresolvedHref: string) {
```

Comment:
prefix private member names with `_`

Reply:
Quoting from the Google TypeScript Style Guide, as I agree with them here :)

```
*_ prefix/suffix*: Identifiers must not use _ as a prefix or suffix. The exception is TypeScript code that is used by JavaScript programmers, e.g. code published on npm. Such code may use _ to mark private properties for outside users not using a compiler.
```

This code is deep within the internals of the analyzer, and I think it's fine not to prefix it with an underscore.

Otherwise every other field and method would have underscores.